File: CTCI/JavaScript/01_Arrays_and_Strings/01_is_unique.js
/**
 * Problem:
 * Implement an algorithm to determine if a string has all unique characters.
 * 
 * Bonus:
 * Implement the algorithm without using a data structure.
 * 
 * @param   {string} str - String to analyze
 * @returns {boolean}      True if all unique characters. False otherwise.
 */

/* Using a data structure */
const hasUniqueChars = (str) => {
  // create new set to store characters
  let charSet = new Set();
  // loop through string and add characters to set
  for (let char of str) {
    // if a character exists in the set, return false
    if (charSet.has(char)) return false;
    charSet.add(char);
  }
  return true;
};

/* Not using a data structure */
const hasUniqueChars = (str) => {
  // split string into array of characters
  str = str.split('');
  // sort character array
  str.sort();
  // loop through character array
  for (let i = 1; i < str.length; i++) {
    // if a character at an index is the same as the last index, return false
    if (str[i] === str[i - 1]) return false;
  }
  return true;
};

File: CTCI/Java/01_Arrays_and_Strings/01_is_unique.java
/**
 * Problem:
 * Implement an algorithm to determine if a string has all unique characters.
 * 
 * Bonus:
 * Implement the algorithm without using a data structure.
 * 
 * @param {string} str String to analyze
 * @return True if all unique characters. False otherwise.
 */

class IsUnique {
    boolean isUniqueChars(String str) {
        if (str.length() > 128)
            return false;

        boolean[] char_set = new boolean[128];
        for (int i = 0; i < str.length(); i++) {
            int val = str.charAt(i);
            // already found this char in string
            if (char_set[val]) {
                return false;
            }
            char_set[val] = true;
        }
        return true;
    }
}